File: interviews/leetcode/leetcode-75/dynamic-programming-multi-dimensional/714. Best Time to Buy and Sell Stock/solution.py
class Solution:
    def maxProfit(self, prices: list[int], fee: int) -> int:
        # Initialize the variables for holding stock and not holding stock
        hold, cash = -prices[0], 0

        for i in range(1, len(prices)):
            # Update the state for holding and not holding the stock
            hold = max(hold, cash - prices[i])
            cash = max(cash, hold + prices[i] - fee)

        return cash

# Main function (commented out by default)
# def main():
#     solution = Solution()
#     prices = [1, 3, 2, 8, 4, 9]
#     fee = 2
#     print(solution.maxProfit(prices, fee))  # Output: 8

File: interviews/leetcode/leetcode-75/dynamic-programming-multi-dimensional/714. Best Time to Buy and Sell Stock/solution.go
package main

// Function to calculate the maximum profit with a transaction fee
func maxProfit(prices []int, fee int) int {
	hold, cash := -prices[0], 0

	for i := 1; i < len(prices); i++ {
		// Update hold and cash for each day
		hold = max(hold, cash-prices[i])
		cash = max(cash, hold+prices[i]-fee)
	}

	return cash
}

// Main function (commented out by default)
// func main() {
//     prices := []int{1, 3, 2, 8, 4, 9}
//     fee := 2
//     fmt.Println(maxProfit(prices, fee)) // Output: 8
// }
